### DISCUSSION TITLE
regarding results in table form and stack bar legends

### DISCUSSION THREAD

**Discussion by Sam131992** [ChiLiubio/microeco#172] [Q&A] [answered]
Dear Microeco Team,
First of all this tool is very powerful and it includes all microbiome-related analysis under one roof which is commendable.
I have a few questions regarding the tool.
1. Is there any option in the tool from which we can extract the analysis data? I mean to get the data in tsv (numeric) form. for example, what per cent of each taxa is present in the stack bar analysis? And in other analyses also can extract similar numeric data?
2. The legends of only 12 taxa in the stacked bar are coloured and the other taxa are showing in the same grey colour. Can we do more Colors of Legends? For example, if we take 20 phyla, then the colours of 12 files look different and 8 of the others have the same colour. Can we show 20 phyla with 20 different colours?

**Comment by ChiLiubio** [comment]
Hi @Sam131992 

Thanks for your questions! These make me know more about the open questions on the package.
1. The easiest way is to save these intermediate data directly according to the description of the 'Return' part of the help document. Please use `help(trans_abund )` and see the `trans_abund$new` Return part of the document. As the trans_abund class call the taxa_abund list in your microtable object, you can also save total abundance files at each taxonomic level using `save_abund` function in microtable class.
2. Please use `color_values` parameter to provide more colors, for example `color_palette_20` inside the package.

Chi

**Reply by Sam131992** [reply]
Thanks for the reply when i tried with color_palette_20 it works but I need more colours in the stack bars so I used randomcoloR but its not working, how to fix this?
2. I am not getting the first query answer..... i am trying to find the numeric value of each analysis but couldn't succeed. please help me in this context.

**Reply by ChiLiubio** [reply]
Try to use it like this `trans_abund$plot_bar(color_values = randomcoloR::randomColor(30))`. I donot know the exact reason without your scripts.
Please see the 'Return' part of the help document in the corresponding class and function. The Return part describes the intermediate data file inside the object. Then you can save it as you want.

**Reply by Sam131992** [reply]
Thank you the color issue is now change but still facing in to find the numeric data.
So I have run this command 
trans_abund$new(
+     dataset = datas,
+     taxrank = "Phylum",
+     show = 0,
+     ntaxa = 40,
+     groupmean = NULL,
+     delete_full_prefix = TRUE,
+     delete_part_prefix = FALSE,
+     prefix = NULL,
+     use_percentage = TRUE,
+     input_taxaname = NULL
+ )
and i got the message 
The transformed abundance data is stored in object$data_abund ...
trans_abund object:
data_abund have 6 columns:  Taxonomy, Sample, Abundance, Fraction, Group, Sample_Name 
data_abund have total 185 taxa
taxrank:  Phylum 
data_taxanames:  Proteobacteria, Actinobacteriota, Planctomycetota, Acidobacteriota, Bacteroidota, Myxococcota, Verrucomicrobiota, Chloroflexota, Firmicutes, Firmicutes_A, Gemmatimonadota, Methylomirabilota, Deinococcota, Thermoproteota, Patescibacteria, Desulfobacterota_B, Desulfobacterota, Halobacteriota, Cyanobacteria, Myxococcota_A, Nitrospirota, Spirochaetota, Armatimonadota, Eisenbacteria, Desulfobacterota_I, Dormibacterota, Elusimicrobiota, Chlamydiota, Eremiobacterota, Bdellovibrionota, Firmicutes_C, Omnitrophota, Firmicutes_B, Hydrogenedentota, Krumholzibacteriota, Thermoplasmatota, Campylobacterota, Firmicutes_G, Latescibacterota, Firmicutes_E 

Now my question is where is the data saved ? i have to write in the manuscript that how much percent of dominant phyla in each group.
Please help me in this regard

**Reply by Sam131992** [reply]
Okay I got it, the abundance files (Phylum, Class, Order, Family, Genus, Species) are saved in the home directory under the trans_abund folder.

**Reply by Sam131992** [reply]
Again one more query can we normalize our data in this tool?

**Reply by ChiLiubio** [reply]
Which way do you want to use to normalize the data? And which data? Theoretically, you can normalize any of the data in the object by accessing it directly.

**Reply by Sam131992** [reply]
I have shotgun sequence data (Taxonomic and Functional) and I want to normalize it by CCS (Cumulative Sum Scaling). Various Research articles show that before analysis we have to normalize the data first. Therefore, I am asking to you is there any function in this tool to normalise the Microbiome data?
Again I am struggling with the detection of the Intermediated data of each analysis. could you please explain it by taking an example? 
Thank you very much for the prompt response it highly appreciable.

**Comment by ChiLiubio** [comment]
Please see trans_norm class of mecodev package for the CCS norm (https://chiliubio.github.io/microeco_tutorial/mecodev-package.html#trans_norm)
An example with trans_abund
```
library(microeco)
data("dataset")
t1 <- trans_abund$new(dataset = dataset, taxrank = "Phylum", ntaxa = 10)
# once you get an object like t1, you can access the object inside it
View(t1$data_abund)
# data_abund is the Return part of trans_abund$new function as the help document conveys in the following picture
# other functions are similar with this
```
![图片](https://user-images.githubusercontent.com/20815519/208298895-65053d04-a8c0-411a-8c78-adf7e5a7088f.png)

**Reply by Sam131992** [reply]
Yes also get the results that " data_abund stored in the object" but where is the object located ?

**Reply by Sam131992** [reply]
Dear Sir,
Following your suggestion I got the table running the command View(t1$data_abund). Please, see the attached picture pf the table. However, how can I transfer this date (Table) into my local system so that my analysis would be easy. 
![Screenshot (16)](https://user-images.githubusercontent.com/116187581/208619990-0b765d7c-0b67-4822-9aea-816e1136f3a4.png)

**Reply by ChiLiubio** [reply]
It is the most commonly-used data.frame format. Please save it using write.table or write.csv function.
```
# something like this
write.csv(t1$data_abund, "test.csv")
# or
write.table(t1$data_abund, "test.tsv", sep = "\t")
```

**Reply by Sam131992** [reply]
Thank you sir it is working.

**Comment by Sam131992** [comment]
Dear Microeco Team,
When I install the mecodev pkagege for normalising my data it shows  
Warning in install.packages :
  package ‘mecodev’ is not available for this version of R

A version of this package for your version of R might be available elsewhere,
see the ideas at
https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages
I am using R 4.2.2 version. how do I rectify this issue?

**Reply by ChiLiubio** [reply]
It is ok to install mecodev online with my R v4.2.2. So please try another way. First download the package and install it locally.
![图片](https://user-images.githubusercontent.com/20815519/208453904-604ea47a-7987-4383-ab73-a0e6dbe05933.png)

**Reply by Sam131992** [reply]
Respected Sir,
by continuous efforts toward learning this tool as well as R. I got all my queries answered. I know I asked lots of silly questions in this discussion, and for that, I apologise. this happened because I am new to R analysis, I have not encountered R analysis. But slowly I am learning this great language. I really appreciate your patience to give the answer promptly to my each silly queries. You are a such great person. Thanks you sir for your help.

**Comment by Sam131992** [comment, marked as answer]
Thank you it's working I have downloaded it.
I found the object in the right corner of the screen (please see the attached picture), my query is how can I download the TSV file which can we see at the left corner of the screen (Highlighted in the line)? I want to import this table into my locale system, I tried many times to import it but cant succeed it. could you help me in this context?
![Screenshot 2022-12-19 at 8 30 12 PM](https://user-images.githubusercontent.com/116187581/208456927-ed9fc461-8c77-4d38-94f8-fe61a684d526.png)

**Reply by ChiLiubio** [reply]
Same way with the upper reply.

**Comment by Sam131992** [comment]
Dear Microeco Team,
I have forget the script for change the colour of taxa in the relative abundance plot.
I have 20 Phylum and 50 taxa but the colour is limited in the default parameter.
Could you add the script in the message if i have 20, 30 and 50 taxa and i want to give a separate colour on each taxon ow can I do that?
t1$plot_bar(color_values = barplot(1:10, col = color_palette_10), facet = c("Group", "Sample_Name"), xtext_keep = FALSE, legend_text_italic = FALSE, barwidth = 1)
I am trying with this but not working.

**Comment by Sam131992** [comment]
Plaese find the below line which is showing everytime 
"Scale for y is already present.
Adding another scale for y, which will replace the existing scale"
what is these lines mean ?
